File: 0084-largest-rectangle-in-histogram/solution.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) 
    {
        const int n = heights.size();
        int maxArea = 0;
        stack<pair<int,int>> S;
        
        for(int i = 0; i < heights.size(); i++)
        {
            int start = i;
            
            while(S.size() && S.top().second > heights[i])
            {
                int index = S.top().first;
                int h     = S.top().second;
                
                S.pop();
                
                maxArea = max(maxArea, h*(i - index));
                start = index;
            }
            S.push({start, heights[i]});
        }
        
        while(S.size())
        {
            int index = S.top().first;
            int h     = S.top().second;

            S.pop();
            
            maxArea = max(maxArea, h*(n - index));

        }
        return maxArea;
        
    }
};

File: 0084-largest-rectangle-in-histogram/solution.py
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        n = len(heights)
        maxArea = 0
        stack = []
        
        for (i, h) in enumerate(heights):
            curIndex = i
            while stack and stack[-1][1] > h:
                index, height = stack.pop()
                maxArea = max(maxArea, height*(i - index))
                curIndex = index

            stack.append((curIndex, h))
        
        while stack:
            index, height = stack.pop()
            maxArea = max(maxArea, height*(n - index))
        
        return maxArea
                
        
        
